query/src/servers/http/v1/query/result_data_manager.rs

```diff
@@ -0,0 +1,147 @@
+use common_datavalues::DataSchemaRef;
+use common_exception::ErrorCode;
+use common_exception::Result;
+
+use crate::servers::http::v1::block_to_json::block_to_json;
+use crate::servers::http::v1::block_to_json::JsonBlock;
+use crate::servers::http::v1::block_to_json::JsonBlockRef;
+
+#[derive(Debug)]
+pub enum Wait {
+    Async,
+    Sync,
+    Deadline(Instant),
+}
+
+#[derive(Clone)]
+pub struct Page {
+    pub data: JsonBlockRef,
+    pub total_rows: usize,
+}
+
+pub struct ResponseData {
+    pub page: Page,
+    pub next_page_no: Option<usize>,
+}
+
+pub struct ResultDataManager {
+    pub(crate) schema: DataSchemaRef,
+    total_rows: usize,
+    total_pages: usize,
+    last_page: Option<Page>,
+    pub(crate) block_rx: mpsc::Receiver<DataBlock>,
+    end: bool,
+}
+
+impl ResultDataManager {
+    pub fn new(schema: DataSchemaRef, block_rx: mpsc::Receiver<DataBlock>) -> ResultDataManager {
+        ResultDataManager {
+            schema,
+            block_rx,
+            total_rows: 0,
+            last_page: None,
+            total_pages: 0,
+            end: false,
+        }
+    }
+
+    pub fn next_page_no(&mut self) -> Option<usize> {
+        if self.end {
+            None
+        } else {
+            Some(self.total_pages)
+        }
+    }
+
+    pub async fn get_a_page(&mut self, page_no: usize, tp: &Wait) -> Result<Page> {
+        let next_no = self.total_pages;
+        if page_no == next_no && !self.end {
+            let (block, end) = self.collect_new_page(tp).await;
+            let num_row = block.len();
+            self.total_rows += num_row;
+            let page = Page {
+                data: Arc::new(block),
+                total_rows: self.total_rows,
+            };
+            if num_row > 0 {
+                self.total_pages += 1;
+                self.last_page = Some(page.clone());
+            }
+            self.end = end;
+            Ok(page)
+        } else if page_no == next_no - 1 {
+            // later, there may be other ways to ack and drop the last page except collect_new_page.
+            // but for now, last_page always exists in this branch, since page_no is unsigned.
+            Ok(self.last_page.as_ref().unwrap().clone())
+        } else {
+            Err(ErrorCode::LogicalError("page no too large"))
+        }
+    }
+
+    pub async fn receive(
+        block_rx: &mut mpsc::Receiver<DataBlock>,
+        tp: &Wait,
+    ) -> std::result::Result<DataBlock, TryRecvError> {
+        use Wait::*;
+        match tp {
+            Async => block_rx.try_recv(),
+            Sync => match block_rx.recv().await {
+                Some(block) => Ok(block),
+                None => Err(TryRecvError::Disconnected),
+            },
+            Deadline(t) => {
+                let sleep = tokio::time::sleep_until(tokio::time::Instant::from_std(*t));
+                tokio::select! {
+                    biased;
+                    block= block_rx.recv() => {
```

Comment:
`block=` ---> `block =` by a `make lint`?

Reply:
ok thanks

Reply:
oh, maybe rustfmt got some trouble on interpreting the macros i guess 🤔